typescript/infra/scripts/verify-validator.ts

```diff
@@ -104,170 +89,259 @@ class S3Wrapper {
   }
 }
 
-async function main() {
-  const {
-    a: _validatorAddress,
-    p: prospectiveBucket,
-    c: controlBucket,
-  } = await getArgs();
+class Validator {
+  private readonly controlS3BucketClient: S3Wrapper;
+  private readonly prospectiveS3BucketClient: S3Wrapper;
 
-  const cClient = new S3Wrapper(controlBucket);
-  const pClient = new S3Wrapper(prospectiveBucket);
+  // accumulators for stats
+  private extraCheckpoints!: number[];
+  private missingCheckpoints!: number[];
+  private invalidCheckpoints!: number[];
+  private modTimeDeltasS!: number[];
+  private fullyCorrectCheckpoints!: number[];
+  private missingInARow!: number;
+  private lastNonMissingCheckpointIndex!: number;
 
-  const [{ obj: controlLatestCheckpoint }, { obj: prospectiveLastCheckpoint }] =
-    await Promise.all([
-      cClient.getS3Obj<number>('checkpoint_latest_index.json').catch((err) => {
-        console.error(
-          "Failed to get control validator's latest checkpoint.",
-          err,
+  constructor(
+    public readonly validatorAddress: string,
+    public readonly controlS3BucketAddress: string,
+    public readonly prospectiveS3BucketAddress: string,
+  ) {
+    this.controlS3BucketClient = new S3Wrapper(this.controlS3BucketAddress);
+    this.prospectiveS3BucketClient = new S3Wrapper(
+      this.prospectiveS3BucketAddress,
+    );
+  }
+
+  initStatsState() {
+    this.extraCheckpoints = [];
+    this.missingCheckpoints = [];
+    this.invalidCheckpoints = [];
+    this.modTimeDeltasS = [];
+    this.fullyCorrectCheckpoints = [];
+    this.missingInARow = 0;
+    this.lastNonMissingCheckpointIndex = Infinity;
+  }
+
+  /**
+   * Validate that the control and prospective validators are in agreement. Will throw an error on
+   * any critical failures and will log stats to the console as it goes.
+   *
+   * If we want to make this callable from outside the script later I would suggest making this
+   * return a stats object or something.
+   */
+  async validate(): Promise<void> {
+    this.initStatsState();
+
+    const { controlLatestCheckpoint, prospectiveLastCheckpoint } =
+      await this.getLatestCheckpoints();
+
+    for (
+      let checkpointIndex = Math.max(
+        controlLatestCheckpoint,
+        prospectiveLastCheckpoint,
+      );
+      checkpointIndex >= 0;
+      --checkpointIndex
+    ) {
+      if (this.missingInARow == MAX_MISSING_CHECKPOINTS) {
```

Comment:
`missingInARow` is not assigned something new before this check in the loop
can we move it to after an update?

Reply:
or should this be part of the `for` loop condition itself? instead of the `break` block

Reply:
I would make it part of the for condition but wanted to have the logic to cleanup if that the condition we ended on. The reason I put it at the beginning of the loop is to address the fact it is basically part of the loop condition, by putting it at the end it would be more difficult to see what the true end conditions are.